File: 77-combinations/77-combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def backtrack(first, curr_combination):
            if len(curr_combination) == k:
                res.append(curr_combination[:])
            
            for i in range(first, n + 1):
                curr_combination.append(i)
                
                backtrack(i + 1, curr_combination)
                
                curr_combination.pop()
                
        res = []
        
        backtrack(1, [])
        return res
        

File: 77-combinations/77-combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    let output = [];
    function backtrack(startIndex, curr_combination) {
        if (curr_combination.length === k) {
           output.push(curr_combination.slice());
        } else {
           for (let i = startIndex; i <= n; i++) {
                curr_combination.push(i)
                backtrack(i + 1, curr_combination);
                curr_combination.pop()
           }
        }
    }
    
    backtrack(1, [])
    return output;
};